### PR TITLE
Cache optimization

### PR DESCRIPTION
**Pull request by JoshuaStorm**
Adds initial cache optimization step with Tipsify. 

See #1 -  Stage for pre and post vertex cache optimization. Use Cesium's implementation.

Any thoughts on making a more specific test for tipsifyGltf without completely copy-pasting the function to get access to the indices array?

~~I added Cesium_Ground.gltf as boxTexturedUnoptimized's indices seem to already be in cache optimized order.~~

### CODE REVIEW

File: lib/tipsifyGltf.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+            var primitivesLength = primitives.length;
+            for (var i = 0; i < primitivesLength; i++) {
+                var primitive = primitives[i];
+                var indexAccessorId = primitive.indices;
+                var indexAccessor = accessors[indexAccessorId];
+                var indicesArray = readAccessor(gltf, indexAccessor);
+                
+                Tipsify.tipsify({
+                    indices: indicesArray.data,
+                    maximumIndex: indexAccessor.max,
```
Comment (lilleyse) [inline lib/tipsifyGltf.js]: `max` is an array so you need to send in the 0th element.

File: lib/tipsifyGltf.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            var mesh = meshes[meshId];
+            var primitives = mesh.primitives;
+            var primitivesLength = primitives.length;
+            for (var i = 0; i < primitivesLength; i++) {
+                var primitive = primitives[i];
+                var indexAccessorId = primitive.indices;
+                var indexAccessor = accessors[indexAccessorId];
+                var indicesArray = readAccessor(gltf, indexAccessor);
+                
+                Tipsify.tipsify({
```
Comment (lilleyse) [inline lib/tipsifyGltf.js]: Just calling the function isn't enough, you'll have to also set the values back to the buffer stored in `extras._pipeline.source`. Maybe you could write a `writeAccessor` type function to handle writing back into the accessor.

File: specs/lib/tipsifyGltfSpec.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+var gltfPath = './specs/data/tipsifyGltf/Cesium_Ground.gltf';
+
+describe('tipsifyGltf', function() {
+    it('reorders indices', function(done) {
+        readGltf(gltfPath, function(gltf) {
+            addDefaults(gltf);
+            var copyGltf = clone(gltf);
+            tipsifyGltf(gltf);
+
+            expect(gltf).not.toEqual(copyGltf);
```
Comment (lilleyse) [inline specs/lib/tipsifyGltfSpec.js]: I was suspicious that this passed, and it looks like it always will. I commented the tipsify line and it passes, meaning clone is always changing something about the gltf. `expect(clone(gltf)).toEqual(copyGltf);` seems to be accurate though, so you should also make that change in `gltfPipelineSpec.js` and anywhere else that does that. For this particular test though, I would just compare the indices alone, not the whole gltf.

File: lib/tipsifyGltf.js (hunk 2 above)
Comment (JoshuaStorm) [inline lib/tipsifyGltf.js]: I was looking at the GeometryPipeline as a guide and it seemed to modify in place, am I missing something here?

Comment (lilleyse) [inline lib/tipsifyGltf.js]: It works a bit differently in Cesium because the indices array sent in is the one that eventually gets used. But in this case the array is just a temporary object and won't be saved with the gltf.

File: specs/lib/tipsifyGltfSpec.js (hunk 3 above)
Comment (JoshuaStorm) [inline specs/lib/tipsifyGltfSpec.js]: I think I'll do this clone fix real quick in a separate PR so it isn't held back by the rest of this PR.

Comment (lilleyse) [inline specs/lib/tipsifyGltfSpec.js]: Ok sounds good.

File: lib/tipsifyGltf.js (hunk 2 above)
Comment (JoshuaStorm) [inline lib/tipsifyGltf.js]: I am currently working on a writeAccessor function, but am at a bit of standstill, which may be caused by an initial misunderstanding of my own:

Currently, I am grabbing the indices from each accessor and tipsifying each set of indices relative to only the indices within that accessor--I am fairly certain this is the way we want to do it.

However, I need to eventually write these tipsified indices to the buffer state, but I cannot do that in the same scope as reading in the accessor, because at that scope I am only looking a specific bufferView and not the entire buffer.

The only solution I am seeing is creating some form of symbol table with keys of the initial byte location and values of the bufferView indices arrays and then writing that to a buffer _after the tipsify loop_. That is a somewhat cumbersome solution, though.

In other words, I'd be doing more of a writeBuffer outside of the loop instead of multiple writeAccessors within the loop since I don't know of a way to just change a single bufferView without rewriting the buffer each time.

Am I missing something? Just wanted to make sure before I go ahead with the cumbersome solution.

I hope I am not missing anything super obvious here. Thanks!

Comment (lilleyse) [inline lib/tipsifyGltf.js]: As we talked about offline, just writing back into the buffer at the correct offset is fine to do.

**Comment by JoshuaStorm** [comment]
Updated. Still need to update tipsifyGltfSpec

**Comment by lilleyse** [comment]
Have you tried running an exported model through the model inspector? It doesn't seem to load correctly, though there are no errors.

**Comment by JoshuaStorm** [comment]
Alright, I think I am good to go for review on this now.

File: lib/gltfPipeline.js [hunk 4]
```diff
@@ -40,19 +41,21 @@ function processJSONWithExtras(gltfWithExtras, options, callback) {
     var stats = new OptimizationStatistics();
 
     addDefaults(gltfWithExtras, stats);
-
+
```
Comment (lilleyse) [inline lib/gltfPipeline.js]: Remove this whitespace.

File: lib/gltfPipeline.js [hunk 5]
```diff
@@ -40,19 +41,21 @@ function processJSONWithExtras(gltfWithExtras, options, callback) {
...
     addDefaults(gltfWithExtras, stats);
-
+    
     removeUnused(gltfWithExtras, stats);
     if (options.printStats) {
         printStats(stats);
     }
     convertDagToTree(gltfWithExtras);
-
+
```
Comment (lilleyse) [inline lib/gltfPipeline.js]: Same

File: lib/writeAccessor.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
+'use strict';
+var byteLengthForComponentType = require('./byteLengthForComponentType');
+var numberOfComponentsForType = require('./numberOfComponentsForType');
+var writeBufferComponentType = require('./writeBufferComponentType');
+
+module.exports = writeAccessor;
+
+function writeAccessor(gltf, accessor, dataArray) { // Argument name?
```
Comment (lilleyse) [inline lib/writeAccessor.js]: The name is fine, you can remove the comment.

File: lib/writeAccessor.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
```
Comment (lilleyse) [inline lib/writeAccessor.js]: Remove empty line.

File: lib/writeAccessor.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    var accessorType = accessor.type;
+    var bufferView = gltf.bufferViews[accessor.bufferView];
+
+    var attributeCount = accessor.count;
+    var componentCount = numberOfComponentsForType(accessorType);
+    var componentType = accessor.componentType;
+    var componentBytes = byteLengthForComponentType(componentType);
+    var currentBufferIndex = accessor.byteOffset + bufferView.byteOffset;
+
+    var buffer = gltf.buffers[bufferView.buffer].extras._pipeline.source;
```
Comment (lilleyse) [inline lib/writeAccessor.js]: For organization purposes just put this under the `bufferView` line.

File: lib/writeAccessor.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+
```
Comment (lilleyse) [inline lib/writeAccessor.js]: Remove extra line.

File: specs/lib/tipsifyGltfSpec.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+'use strict';
+var tipsifyGltf = require('../../lib/tipsifyGltf');
+var addDefaults = require('../../lib/addDefaults');
+var readGltf = require('../../lib/readGltf');
+var writeAccessor = require('../../lib/writeAccessor');
```
Comment (lilleyse) [inline specs/lib/tipsifyGltfSpec.js]: This include isn't used.

File: specs/lib/tipsifyGltfSpec.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+var readAccessor = require('../../lib/readAccessor');
+
+var gltfPath = './specs/data/tipsifyGltf/chinesedragon.gltf';
+
+describe('tipsifyGltf', function() {
+    it('reorders indices', function(done) {
+        readGltf(gltfPath, function(gltf) {
+            addDefaults(gltf);
+            var indexAccessorId = gltf.meshes[Object.keys(gltf.meshes)].primitives[0].indices;
```
Comment (lilleyse) [inline specs/lib/tipsifyGltfSpec.js]: Should be `Object.keys(gltf.meshes)[0]`.

**Comment by lilleyse** [comment]
I'm starting to think that `tipsifyGltf.js` should be renamed to be more descriptive of the stage, like `cacheOptimizeIndices.js` or something.

**Comment by JoshuaStorm** [comment]
Referencing back to #1 

"Stage for _pre_ and _post_ vertex cache optimization. Use Cesium's implementation."

Should cache optimization be applied prior to and at the end of the pipeline?

**Comment by lilleyse** [comment]
The pre/post terminology is just describing the technique used by the gpu. (https://www.opengl.org/discussion_boards/showthread.php/167481-Vertex-cache-optimization). Only doing it once in our pipeline is fine.

File: lib/cacheOptimizeIndices.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+            var primitivesLength = primitives.length;
+            for (var i = 0; i < primitivesLength; i++) {
+                var primitive = primitives[i];
+                var indexAccessorId = primitive.indices;
+                var indexAccessor = accessors[indexAccessorId];
+                var indicesArray = readAccessor(gltf, indexAccessor);
+
+                indicesArray.data = Tipsify.tipsify({
+                    indices: indicesArray.data,
+                    maximumIndex: undefined,
```
Comment (lilleyse) [inline lib/cacheOptimizeIndices.js]: My comment may have gotten lost, but you can remove this line.

**Comment by lilleyse** [comment]
Make sure to also change the spec name and the spec/data folder name.

**Comment by lilleyse** [comment]
Looks good!

File: specs/lib/cacheOptimizeIndicesSpec.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+var cacheOptimizeIndices = require('../../lib/cacheOptimizeIndices');
+var addDefaults = require('../../lib/addDefaults');
+var readGltf = require('../../lib/readGltf');
+var readAccessor = require('../../lib/readAccessor');
+
+var gltfPath = './specs/data/cacheOptimizeIndices/chinesedragon.gltf';
```
Comment (pjcozzi) [inline specs/lib/cacheOptimizeIndicesSpec.js]: How many triangles is this model?  To keep the repo small and the tests fast, why not just run this on a box?  The Cesium function is already unit tested, we just need to test the integration.

File: lib/cacheOptimizeIndices.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+            var primitivesLength = primitives.length;
+            for (var i = 0; i < primitivesLength; i++) {
+                var primitive = primitives[i];
+                var indexAccessorId = primitive.indices;
+                var indexAccessor = accessors[indexAccessorId];
+                var indicesArray = readAccessor(gltf, indexAccessor);
+
+                indicesArray.data = Tipsify.tipsify({
+                    indices: indicesArray.data,
+                    cacheSize: 24
```
Comment (pjcozzi) [inline lib/cacheOptimizeIndices.js]: This should be a parameter that we expose.

File: specs/lib/cacheOptimizeIndicesSpec.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    it('reorders indices', function(done) {
+        readGltf(gltfPath, function(gltf) {
+            addDefaults(gltf);
+            var indexAccessorId = gltf.meshes[Object.keys(gltf.meshes)[0]].primitives[0].indices;
+            var indicesBefore = readAccessor(gltf, gltf.accessors[indexAccessorId]);
+            
+            cacheOptimizeIndices(gltf);
+            
+            var indicesAfter = readAccessor(gltf, gltf.accessors[indexAccessorId]);
+            expect(indicesBefore).not.toEqual(indicesAfter);
```
Comment (pjcozzi) [inline specs/lib/cacheOptimizeIndicesSpec.js]: This should check for specific values, and should also check vertices since they may change for the pre-vertex shader cache.

File: specs/lib/cacheOptimizeIndicesSpec.js (hunk 13 above)
Comment (JoshuaStorm) [inline specs/lib/cacheOptimizeIndicesSpec.js]: The cache optimization seems to have no effect on the usual box we use for testing -- I assumed this was reasonable since there are so few indices. Let me see if I can figure out how to reorder the indices such that they will be moved in a known manner with the cache optimization so I can avoid using the dragon model.

**Comment by pjcozzi** [comment]
> The cache optimization seems to have no effect on the usual box we use for testing -- I assumed this was reasonable since there are so few indices. Let me see if I can figure out how to reorder the indices such that they will be moved in a known manner with the cache optimization so I can avoid using the dragon model.

Or just try another simple test model if we have one.

File: specs/lib/cacheOptimizeIndicesSpec.js (hunk 15 above)
Comment (JoshuaStorm) [inline specs/lib/cacheOptimizeIndicesSpec.js]: > should also check vertices since they may change for the pre-vertex shader cache.

Right now I believe I am only doing the post vertex cache optimization with tipsify. Another pull request will implement the pre vertex cache optimization
